components/places/src/storage/bookmarks.rs

```diff
@@ -0,0 +1,1101 @@
+    pub title: Option<String>,
+    pub children: Vec<BookmarkTreeNode>,
+}
+
+#[cfg(test)]
+impl PartialEq for FolderNode {
+    fn eq(&self, other: &FolderNode) -> bool {
+        cmp_option!(self, other, guid)
+            && cmp_option!(self, other, date_added)
+            && cmp_option!(self, other, last_modified)
+            && cmp_option!(self, other, title)
+            && self.children == other.children
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum BookmarkTreeNode {
+    Bookmark(BookmarkNode),
+    Separator(SeparatorNode),
+    Folder(FolderNode),
+}
+
+// Serde makes it tricky to serialize what we need here - a 'type' from the
+// enum and then a flattened variant struct. So we gotta do it manually.
+impl Serialize for BookmarkTreeNode {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("BookmarkTreeNode", 2)?;
+        match self {
+            BookmarkTreeNode::Bookmark(b) => {
+                state.serialize_field("type", &BookmarkType::Bookmark)?;
+                state.serialize_field("guid", &b.guid)?;
+                state.serialize_field("date_added", &b.date_added)?;
+                state.serialize_field("last_modified", &b.last_modified)?;
+                state.serialize_field("title", &b.title)?;
+                state.serialize_field("url", &b.url.to_string())?;
+            }
+            BookmarkTreeNode::Separator(s) => {
+                state.serialize_field("type", &BookmarkType::Separator)?;
+                state.serialize_field("guid", &s.guid)?;
+                state.serialize_field("date_added", &s.date_added)?;
+                state.serialize_field("last_modified", &s.last_modified)?;
+            }
+            BookmarkTreeNode::Folder(f) => {
+                state.serialize_field("type", &BookmarkType::Folder)?;
+                state.serialize_field("guid", &f.guid)?;
+                state.serialize_field("date_added", &f.date_added)?;
+                state.serialize_field("last_modified", &f.last_modified)?;
+                state.serialize_field("title", &f.title)?;
+                state.serialize_field("children", &f.children)?;
+            }
+        };
+        state.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for BookmarkTreeNode {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // *sob* - a union of fields we post-process.
+        #[derive(Debug, Default, Deserialize)]
+        #[serde(default)]
+        struct Mapping {
+            #[serde(rename = "type")]
+            bookmark_type: u8,
+            guid: Option<SyncGuid>,
+            date_added: Option<Timestamp>,
+            last_modified: Option<Timestamp>,
+            title: Option<String>,
+            #[serde(with = "url_serde")]
+            url: Option<Url>,
```

Comment:
We had issues with using url_serde here in https://github.com/mozilla/application-services/issues/571 :(